documentation/general/analyzer-redirecting.md

```diff
@@ -0,0 +1,129 @@
+# Redirecting analyzers in SDK to VS
+
+We will redirect analyzers from the SDK to ones deployed in the VS to avoid the [torn SDK][torn-sdk] issue.
+Only major versions will be redirected because different major versions of the same analyzer are not compatible.
+So this applies to a situation like:
+- Having an analyzer in SDK 9.0.1 referencing Roslyn 4.12. That gets deployed to VS 17.12.
+- Having an analyzer in SDK 9.0.7 referencing Roslyn 4.13.
+- User loads project with SDK 9.0.7 in VS 17.12.
+  Previously the analyzers would fail to load as they reference newer Roslyn version (4.13) than what is part of VS (4.12).
+  Now we will redirect the analyzers to those from SDK 9.0.1 that are deployed as part of VS and reference the matching Roslyn (4.12).
+
+Loading analyzers with older major version should not be a problem because they must reference an older version of Roslyn.
+
+Loading analyzers with newer major version in an old VS will result in an error like:
+
+> error NETSDK1045: The current .NET SDK does not support targeting .NET 10.0.  Either target .NET 9.0 or lower, or use a version of the .NET SDK that supports .NET 10.0. Download the .NET SDK from
```

Comment:
Confused by this error. This is the error about TFM targetting while analyzer load errors are generally about the compiler version.

Reply:
If you try to load a new .NET SDK in an old VS, you will get this error even before the analyzers are loaded, I think.

Reply:
Think that is only true if you also target the new TFM.

Reply:
That makes sense. So if a user has SDK 10 (with analyzers 10.\*), a project targeting .NET 9, and VS 17.12 (with inserted analyzers 9.\*), we won't redirect the analyzers because major versions don't match, and they will fail with the roslyn mismatch errors.

I don't understand how would this work in command-line builds though. If the SDK 10 only contains 10.\* analyzers, and a project targets .NET 9, we shouldn't use the 10.\* analyzers because they might not be compatible with the 9.\* analyzers, right?

Reply:
Looks like command-line builds use the older pack or download it if it doesn't exist, but they still load some analyzers from the current version:

![image](https://github.com/user-attachments/assets/41133f83-9df2-40d5-8797-2a7af8bd1082)

But in VS we won't have the analyzers from the older pack nor we can load the current analyzers because they might be incompatible. So I guess we will just error.

Reply:
`dotnet new console` with net 9 preview 6 with tfm net8.0